pkg/coderef/storage/v2/code_reference.go

```diff
@@ -0,0 +1,232 @@
+		cr.RepositoryName,
+		cr.RepositoryOwner,
+		cr.RepositoryType,
+		cr.RepositoryBranch,
+		cr.CommitHash,
+		cr.UpdatedAt,
+		cr.Id,
+		cr.EnvironmentId,
+	)
+	if err != nil {
+		return err
+	}
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if rowsAffected != 1 {
+		return ErrCodeReferenceUnexpectedAffectedRows
+	}
+	return nil
+}
+
+func (s *codeReferenceStorage) GetCodeReference(
+	ctx context.Context,
+	id, environmentID string,
+) (*domain.CodeReference, error) {
+	codeRef := &domain.CodeReference{}
+	row := s.qe(ctx).QueryRowContext(
+		ctx,
+		selectCodeReferenceSQL,
+		id,
+		environmentID,
+	)
+	err := row.Scan(
+		&codeRef.Id,
+		&codeRef.FeatureId,
+		&codeRef.FilePath,
+		&codeRef.LineNumber,
+		&codeRef.CodeSnippet,
+		&codeRef.ContentHash,
+		&mysql.JSONObject{Val: &codeRef.Aliases},
+		&codeRef.RepositoryName,
+		&codeRef.RepositoryOwner,
+		&codeRef.RepositoryType,
+		&codeRef.RepositoryBranch,
+		&codeRef.CommitHash,
+		&codeRef.EnvironmentId,
+		&codeRef.CreatedAt,
+		&codeRef.UpdatedAt,
+	)
+	if err != nil {
+		if errors.Is(err, mysql.ErrNoRows) {
+			return nil, ErrCodeReferenceNotFound
+		}
+		return nil, err
+	}
+	return codeRef, nil
+}
+
+func (s *codeReferenceStorage) ListCodeReferences(
+	ctx context.Context,
+	whereParts []mysql.WherePart,
+	orders []*mysql.Order,
+	limit, offset int,
+) ([]*domain.CodeReference, int, int64, error) {
+	whereSQL, whereArgs := mysql.ConstructWhereSQLString(whereParts)
+	orderBySQL := mysql.ConstructOrderBySQLString(orders)
+	limitOffsetSQL := mysql.ConstructLimitOffsetSQLString(limit, offset)
+	query := fmt.Sprintf("%s %s %s %s", selectCodeReferencesSQL, whereSQL, orderBySQL, limitOffsetSQL)
+	rows, err := s.qe(ctx).QueryContext(ctx, query, whereArgs...)
+	if err != nil {
+		return nil, 0, 0, err
+	}
+	defer rows.Close()
+	codeRefs := make([]*domain.CodeReference, 0, limit)
+	for rows.Next() {
+		codeRef := &domain.CodeReference{}
+		err := rows.Scan(
+			&codeRef.Id,
+			&codeRef.FeatureId,
+			&codeRef.FilePath,
+			&codeRef.LineNumber,
+			&codeRef.CodeSnippet,
+			&codeRef.ContentHash,
+			&mysql.JSONObject{Val: &codeRef.Aliases},
+			&codeRef.RepositoryName,
+			&codeRef.RepositoryOwner,
+			&codeRef.RepositoryType,
+			&codeRef.RepositoryBranch,
+			&codeRef.CommitHash,
+			&codeRef.EnvironmentId,
+			&codeRef.CreatedAt,
+			&codeRef.UpdatedAt,
+		)
+		if err != nil {
+			return nil, 0, 0, err
+		}
+		codeRefs = append(codeRefs, codeRef)
+	}
+	if rows.Err() != nil {
+		return nil, 0, 0, rows.Err()
+	}
+	nextOffset := offset + len(codeRefs)
+	var total int64
+	countQuery := fmt.Sprintf("%s %s", countCodeReferencesSQL, whereSQL)
+	row := s.qe(ctx).QueryRowContext(ctx, countQuery, whereArgs...)
+	if err := row.Scan(&total); err != nil {
+		return nil, 0, 0, err
+	}
+	return codeRefs, nextOffset, total, nil
+}
+
+func (s *codeReferenceStorage) DeleteCodeReference(ctx context.Context, id, environmentID string) error {
```

Comment:
Since we are using UUID, we don't need the environment ID in the storage layer.

```suggestion
func (s *codeReferenceStorage) DeleteCodeReference(ctx context.Context, id string) error {
```